### PR TITLE
KAA-331: Implemented Facebook verifier 

### PR DESCRIPTION
**Pull request by batytskyy**

### CODE REVIEW

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 1]
```diff
@@ -67,47 +69,20 @@ public TokenVerifier(String userExternalId, String userAccessToken,
...
-                    } else if (!receivedUserId.equals(userExternalId)) {
-                        LOG.trace("Input token doesn't belong to the user with {} id", userExternalId);
-                        callback.onVerificationFailure("User access token " + userAccessToken + " doesn't belong to the user");
-                    } else {
-                        LOG.trace("Input token is confirmed and belongs to the user with {} id", userExternalId);
-                        callback.onSuccess();
-                    }
-                } else {
+
+                if (connection.getResponseCode() == 400) {              // HTTP 400: Bad request
```
Comment (ashvayka) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: Move "400" to constant

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 2]
```diff
@@ -67,47 +69,20 @@ public TokenVerifier(String userExternalId, String userAccessToken,
...
+                    handle400Response(connection, userExternalId, callback, userAccessToken);
```
Comment (ashvayka) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: handleBadRequest

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 3]
```diff
@@ -67,47 +69,20 @@ public TokenVerifier(String userExternalId, String userAccessToken,
...
+                } else if (connection.getResponseCode() == 200) {       // HTTP 200: OK
```
Comment (ashvayka) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: Move "200" to constant

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 4]
```diff
@@ -67,47 +69,20 @@ public TokenVerifier(String userExternalId, String userAccessToken,
...
+                    handle200Response(connection, userExternalId, callback, userAccessToken);
```
Comment (ashvayka) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: handleResponse

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 5]
```diff
@@ -37,9 +37,11 @@
     private static final Logger LOG = LoggerFactory.getLogger(FacebookUserVerifier.class);
     private static final String FACEBOOK_URL_PREFIX = "https://graph.facebook.com/debug_token";
```
Comment (shpigor) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: I think its better move url to properties

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 6]
```diff
@@ -124,12 +99,80 @@ public void run() {
...
+
+            Map<String, Object> dataMap = (Map<String, Object>) responseMap.get("data");
+            String receivedUserId = String.valueOf(dataMap.get("user_id"));
+
+            if (dataMap.containsKey("error") || receivedUserId == null) {
+                Map<String, Object> errorMap = (Map<String, Object>) dataMap.get("error");
+                LOG.trace("Bad input token: {}, errcode = {}", errorMap.get("message"), errorMap.get("code"));
+                callback.onTokenInvalid();
+            } else if (!receivedUserId.equals(userExternalId)) {
+                LOG.trace("Input token doesn't belong to the user with {} id", userExternalId);
```
Comment (shpigor) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: change to warn log level

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 7]
```diff
@@ -124,12 +99,80 @@ public void run() {
...
+            // we always get a map
+            Map<String, Object> responseMap =
+                    responseMapper.readValue(reader.readLine(), Map.class);
+
+            Map<String, Object> dataMap = (Map<String, Object>) responseMap.get("data");
+            String receivedUserId = String.valueOf(dataMap.get("user_id"));
+
+            if (dataMap.containsKey("error") || receivedUserId == null) {
+                Map<String, Object> errorMap = (Map<String, Object>) dataMap.get("error");
+                LOG.trace("Bad input token: {}, errcode = {}", errorMap.get("message"), errorMap.get("code"));
```
Comment (shpigor) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: change to warn or error log level

File: server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java [hunk 8]
```diff
@@ -124,12 +99,80 @@ public void run() {
...
+
+            // no error field in response
+            if (responseMap.get("error") != null) {
+                errorMap = (Map<String, Object>) responseMap.get("error");
+            }
+
+            // errors with OAuth
+            if (errorMap != null && String.valueOf(errorMap.get("code")).equals(OAUTH_ERROR)) {
+                if (errorMap.get("error_subcode") == null) {
+                    LOG.trace("OAuthException: [{}], errcode: [{}], errsubcode: [{}] ", errorMap.get("message"),
```
Comment (shpigor) [inline server/verifiers/facebook-verifier/verifier/src/main/java/org/kaaproject/kaa/server/verifiers/facebook/verifier/FacebookUserVerifier.java]: Please move to constants string variables like "message", "errcode" ... etc.

**Comment by shpigor** [comment]
Reviewed

**Pull request merged**